### ISSUE TITLE
Beacon implementation fails in Chrome and Safari 12+ due to browser bugs

### ISSUE THREAD

**Issue by mgrahamx** [snowplow/snowplow-javascript-tracker#751]
Hey guys,

The implementation of Beacon API support in this library appears to be impacted by two separate browser bugs.

https://github.com/snowplow/snowplow-javascript-tracker/blob/ed751aaf8486f72d0abc0f0f880754c3dd61aa65/src/js/out_queue.js#L306

Here you are sending a Beacon request with JSON as the type, which is not in Chrome's CORS pre-approved headers list. This causes Chrome to throw an exception when you try to send a beacon this way. Of course this means it will fallback to POST and correctly log the event.

ref: https://bugs.chromium.org/p/chromium/issues/detail?id=490015

In Safari 12 & 13, it appears the same bug is present however it will silently fail and send a blank request body. This is a much bigger problem because it doesn't fallback and no tracking data is actually sent.

**Comment by mgrahamx** [comment]
Have not tested Firefox, but I suspect the underlying security vulnerability that causes this bug has either been fixed or it throws similarly to Chrome. We aren't seeing Firefox tracking issues.

**Comment by phliem** [comment]
I had the exact same problem. On Chrome 77.0 and Safari 13.0.3. Firefox was working perfectly though. The events were sent properly to snowplow.

Not sure if it is the best way to fix this, but changing the `eventMethod` to no use Beacon works for me. Just in case someone else needs to fix the events on chrome too here what I did.

On initialisation in the argmap, change the `eventMethod` to `get` or `post` 
Here is an example
```
snowplow("newTracker", "WebTracker", snowplowCollectorUrl, {
    appId: "Webapp",
    platform: "web",
    eventMethod: "post"
});
```

**Comment by paulboocock** [comment]
We've done some investigation into this and we can confirm it's an issue effecting browsers based on Chromium (Chrome, Opera, Edge Beta) however events are still sent using POST. 

This issue doesn't effect Firefox as it sends a preflight and then correctly sends the request.

As the Chromium based browsers throw an exception, which we are handling, we fall back to POST requests and events are then sent. There isn't much we can do until the Chromium team improve their handling of beacon requests, we're are now tracking https://bugs.chromium.org/p/chromium/issues/detail?id=724929 which should fix this issue in Chromium.

As for Safari, I think this warrants some further investigation. Safari's beacon implementation appears a little odd at first glance but it does appear to work on my initial test requests but it doesn't seem to send a preflight request as I would have expected for a cross domain request.

**Comment by ScottyMJacobson** [comment]
@mgrahamx good catch - for more context on this and a potential workaround - look at the conversation on #704. our workaround now is that our collector actually now just accepts `text/plain`, and we send out our beacon without making the payload a blob with the application/json headers (even though it really is `application/json`)

#hacksonhacks... this is reminding me to check the Chromium tickets for any updates

**Comment by paulboocock** [comment]
@ScottyMJacobson in _slightly_ exciting news, the chromium team marked this as fixed yesterday https://bugs.chromium.org/p/chromium/issues/detail?id=724929
Hopefully this will allow Chromium based browsers to start sending events via Beacon rather than falling back to Post.

**Comment by ScottyMJacobson** [comment]
@paulboocock 😮 file that under "Things I Didn't Expect to Hear on a Thursday Morning"

[how's that for serendipity?](https://en.wikipedia.org/wiki/Synchronicity)

if i remember correctly, it takes chrome like 2-3 months to go from commit->canary->release, right?

**Comment by paulboocock** [comment]
Yeah that sounds about right, although there is Dev and Beta between Canary and Release. I downloaded Canary today to see if it was too good to be true but sure enough, the beacon events in Chrome no longer throw an exception and successfully end up in my collector. The only oddity is the type shows as text/plain but I'm hoping that is a display issue :)
![Screenshot 2020-01-23 at 17 13 10](https://user-images.githubusercontent.com/8820009/73007047-a3abf100-3e03-11ea-8e81-3740619223d9.png)

**Comment by paulboocock** [comment]
Closing as Chromium fix is incoming and we'e just released 2.13.0 which includes better Safari support (https://github.com/snowplow/snowplow-javascript-tracker/issues/716)
